modelling/feat_engineer/feature_engineering.py

```diff
@@ -0,0 +1,300 @@
+    ln_vp_940 = np.log(vp_940)
+
+    print(ln_vp_600, ln_vp_940)
+    print(m_600, m_940)
+    print(hr_600, hr_940)
+
+    spo2 = (e_hb_600 * np.sqrt(m_940 * ln_vp_940) - e_hb_940 * np.sqrt(m_600 * ln_vp_600)) / \
+        (np.sqrt(m_940 * ln_vp_940)*(e_hb_600 - e_hbo_600) -
+         np.sqrt(m_600 * ln_vp_600)*(e_hb_940 - e_hbo_940))
+
+    return spo2.mean(), spo2.std()
+
+
+def _wavelet_filter_signal(s, wave='db4', *args, **kwargs):
+    cA, cD = pywt.dwt(s, wave)
+    cD_mod = pywt.threshold(cD, *args, **kwargs)
+    s_mod = pywt.idwt(cA, cD_mod, wave)
+    return s_mod
+
+
+def rgb_to_ppg(df: pd.DataFrame, filter='band', qe_constants=(0.004989, 0.001193, 0.001), scale=False, block_id='sample_id') -> pd.DataFrame:
+    """Turn RGB means into PPG curves as per Lamonaca.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        Dataframe with the colour data per frame (`mean_red`, etc.)
+    filter : str, optional
+        The kind of signal filter to apply, by default 'band'.
+        `band` = bandpass Butterworth filter
+        `low` = lowpass Butterworth filter
+        `firwin` = lowpass Firwin filter
+    qe_constants : (float, float, float)
+        The Quantum Efficiency constants (Lamonaca et al.) to use for scaling the
+        means of the (red, green, blue) channels respectively.
+    block_id : str, optional
+        The field to use for chunking the data before applying transforms,
+        by default 'sample_id'.
+
+    Returns
+    -------
+    pd.DataFrame
+        Copy of the original dataframe, but with 'tx_red', 'tx_blue', 'tx_green'.
+
+    """
+
+    if filter not in ['band', 'low', 'firwin', None]:
+        raise ValueError("filter must be 'band', 'low', 'firwin', or None")
+
+    _df = df.copy()
+
+    for i, colour in enumerate(['red', 'green', 'blue']):
+        qe_const = qe_constants[i]
+        _df[f'tx_{colour}'] = _df[f'mean_{colour}'] * qe_const
+
+    tx_fields = ['tx_red', 'tx_green', 'tx_blue']
+
+    '''Filtering'''
+    fs = 30.0       # sample rate, Hz
+    nyq = 0.5 * fs  # Nyquist Frequency
+    order = 2
+
+    if filter is not None:
+        if filter == 'band':
+            low = 0.3 / nyq
+            high = 4.2 / nyq
+            b, a = butter(order, [low, high], btype='band', analog=False)
+        else:
+            cutoff = 4.0 / nyq    # desired cutoff frequency of the filter, Hz
+            if filter == 'low':
+                b, a = butter(order, cutoff, btype='low', analog=False)
+            if filter == 'firwin':
+                b = firwin(order+1, cutoff, )
+                a = 1.0
+
+        # Apply the filter (and normalise)
+        for bid in _df[block_id].unique():
+            for field in tx_fields:
+                _df.loc[_df[block_id] == bid, field] = filtfilt(
+                    b, a, _df[_df[block_id] == bid][field])
+
+                # Scaling
+                if scale:
+                    _df.loc[_df[block_id] == bid, field] = MinMaxScaler(
```

Comment:
Maybe some artifact detection should be applied first. If we normalize, the highly noisy outliers would overpower the actual signal.

Reply:
That's a good point. The band-pass filter should clean up a fair amount of that, but I agree that having some additional outlier detection would help. I've opened [an issue](https://github.com/gianlucatruda/Spo2_evaluation/issues/3) for this